crypto/sha/asm/sha512-x86_64.pl

```diff
@@ -1992,7 +1992,23 @@ ()
 	vmovdqa	$t0,0x00(%rsp)
 	xor	$a1,$a1
 	vmovdqa	$t1,0x20(%rsp)
+___
+$code.=<<___ if (!$win64);
+# temporarily use %rdi as frame pointer
+	mov	$_rsp,%rdi
+.cfi_def_cfa	%rdi,8
+___
+$code.=<<___;
 	lea	-$PUSH8(%rsp),%rsp
+___
+$code.=<<___ if (!$win64);
+# the frame info is at $_rsp, but the stack is moving...
+# so a second frame pointer is saved at -8(%rsp)
+# that is in the red zone
```

Comment:
the red zone is a stack area from %rsp-128..%rsp-1 that is preserved on x84_64 linux systems at least
it is not available on windows, therefore most of the extra code is guarded by if (!$win64).
All this stack trace code here is likely not working on windows.

Reply:
> All this stack trace code here is likely not working on windows anyway.

I haven't looked at this PR (using the red zone seems like a bit of a hack), but OpenSSL's assembly is far from ABI-compliant on Windows. Win64 mandates unwind information for all functions and it is much much more restrictive than CFI. The handwritten SEH handlers aren't actually how you're supposed to do this.

Fixing this will require changing the assembly files a bit to retain proper frame pointers and such (though not everything needs a frame pointer, only functions with variable stack allocations). I have some work I need to finish up that implements perlasm directives to emit Win64 unwind codes, as well as an automated instruction-by-instruction tester for both Windows and Linux unwind in BoringSSL.

Reply:
According to the System V Application Binary Interface
see https://software.intel.com/sites/default/files/article/402129/mpx-linux64-abi.pdf:
"The 128-byte area beyond the location pointed to by %rsp is considered to be reserved and shall not be modified by signal or interrupt handlers. Therefore, functions may use this area for temporary data that is not needed across function calls.  In particular, leaf functions may use this area for their entire stack frame, rather than adjusting the stack pointer in the prologue and epilogue.  This area is known as the red zone."
So that seems to be correct to use.
I have no idea how to fix windows.
What is clear that this PR does not attempt to store the frame pointer in the red zone
for windows, since I doubt anyway the cfi info will work as is and I am unable to test
anything reasonably under windows.